### PR TITLE
Messaging client

### LINKED ISSUE #5: Implement interface to message queue.
Kafka will be used as a message queue. A messaging interface must allow clients to publish add, delete, and update operations.

### PR DESCRIPTION
**Pull request by nivista**
closes #5

### CODE REVIEW

File: messaging/client.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+	producer   sarama.SyncProducer
+	partitions int
+	topic      string
+}
+
+// NewClient returns a new Client.
+func NewClient() *Client {
+	client := Client{}
+
+	client.topic = os.Getenv("TIMER_TOPIC")
```
Comment (andrewjdawson2016) [inline messaging/client.go:25]: We need some config files

File: messaging/client.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+	partitions, err := strconv.Atoi(os.Getenv("TIMER_PARTITIONS"))
+	if err != nil {
+		panic(err)
```
Comment (andrewjdawson2016) [inline messaging/client.go:29]: why do we panic here but return an error for database client?

File: messaging/client.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+	}
+	client.partitions = partitions
+
+	prodConfig := sarama.NewConfig()
+	prodConfig.Producer.RequiredAcks = sarama.WaitForAll
```
Comment (andrewjdawson2016) [inline messaging/client.go:34]: How do we configure log compaction?

Comment (andrewjdawson2016) [inline messaging/client.go:34]: Why wait for all instead of quorm?

File: messaging/client.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+// PublishDelete publishes a timer deletion to the queue synchronously.
+func (c *Client) PublishDelete(domain string, id uuid.UUID) error {
+	key, err := utils.NewTimerKey(domain, id)
+	if err != nil {
+		return err
+	}
+
+	partition := c.idToPartition(id)
+
+	m := sarama.ProducerMessage{
```
Comment (andrewjdawson2016) [inline messaging/client.go:67]: We need a type field here to switch on.

File: messaging/client.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+		Topic:     c.topic,
```
Comment (andrewjdawson2016) [inline messaging/client.go:68]: I think we should be using proto from day one for these.

File: messaging/client.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+		Key:       key,
+		Value:     sarama.ByteEncoder(timerBytes),
+		Partition: partition,
+	}
+
+	_, _, err = c.producer.SendMessage(&m)
+	return err
+}
+
+func (c *Client) idToPartition(id [16]byte) int32 {
```
Comment (andrewjdawson2016) [inline messaging/client.go:100]: use a hash library to get. some hashfingerprint here.

**Review by andrewjdawson2016 — CHANGES_REQUESTED** [review]
We have some work to do :)

File: messaging/client.go (hunk 3 above)
Comment (nivista) [reply messaging/client.go:34]: Wait for all is wait for all in sync replicas, Kafka doesn't do quorums

Comment (nivista) [reply messaging/client.go:34]: That's done on the server side of Kafka. Should Kafka configuration be put in version control somehow?

File: messaging/client.go (hunk 2 above)
Comment (nivista) [reply messaging/client.go:29]: Ya that is a good point. I'll return error in both places.

File: messaging/client.go (hunk 1 above)
Comment (nivista) [reply messaging/client.go:25]: Ok. And I should push these to version control?

File: messaging/client.go (hunk 6 above)
Comment (nivista) [reply messaging/client.go:100]: ok.

File: messaging/client.go (hunk 5 above)
Comment (nivista) [reply messaging/client.go:68]: ok.

File: messaging/client.go (hunk 4 above)
Comment (nivista) [reply messaging/client.go:67]: What do you mean?

**Pull request closed without merging**